amp-pwa-reader/src/server/server.js

```diff
@@ -0,0 +1,165 @@
+    }
+  });
+});
+
+
+// When user requests an article, serve the AMP version of that article,
+// injecting our service worker and replacing the Guardian's menu with one that works for Shadow Reader.
+app.get('/' + pub.pathname + '/*', (req, res) => {
+  let isSectionUrl = !req.params[0].match(/\/./);
+
+// If this URL is for a section, not an article, then return the Shadow Reader instead.
+  if (isSectionUrl) {
+    res.sendFile(path.join(__dirname, '../index.html'));
+
+  } else {
+
+// Build the AMP URL from the Shadow Reader URL, and request the AMP.
+    let categoryAndUrl = splitUrlCategory(req.params[0]);
+    const ampUrl = pub.constructAMPUrl(categoryAndUrl.category, categoryAndUrl.url);
+
+    request(ampUrl, function(error, response, body) {
+      if (!error) {
+        res.send(addServiceWorker(replaceSidebar(body)));
+      } else {
+        res.json({error: 'An error occurred in the article route'});
+      }      
+    });
+
+  }
+});
+
+
+// listen for requests :)
+const port = process.env.PORT || 8080;
+const listener = app.listen(port, () => {
+  console.log('Your app is listening on port ' + listener.address().port);
+});
+
+
+
+/** HELPERS **/
+
+// Cache with memory-cache. A tip of the pin to https://medium.com/the-node-js-collection/simple-server-side-cache-for-express-js-with-node-js-45ff296ca0f0
+// Essentially, we wrap res.send() in a function that first puts the result in the cache.
+function cache(seconds) {
+  return (req, res, next) => {
+    const key = req.originalUrl || req.url;
+    const cachedBody = memCache.get(key);
+
+    if (cachedBody) {
+      console.log('** cache hit: ' + key);
+      res.send(cachedBody);
+
+    } else {
+      console.log('** cache miss: ' + key);
+      res.justSend = res.send;
+      res.send = (body) => {
+        memCache.put(key, body, seconds * 1000);
+        res.justSend(body);
+      };
+
+      next();
+    }
+  }
+}
+
+// Inject service worker HTML into an HTML document.
+function addServiceWorker(html) {
+  const swHeadHTML = fs.readFileSync(path.join(__dirname, 'partials/sw_head.html'), 'utf8');
+  const swBodyHTML = fs.readFileSync(path.join(__dirname, 'partials/sw_body.html'), 'utf8');
+
+  return html.replace('</head>', swHeadHTML + "\n$&")
+             .replace(/<body.+?>/, "$&\n" + swBodyHTML)
+  ;
+}
+
+// Replace the sidebar with one that works for the Shadow Reader.
+function replaceSidebar(html) {
+  const template = fs.readFileSync(path.join(__dirname, 'partials/sidebar.html'), 'utf8');
+  const basePath = '/' + pub.pathname + '/';
+  let newSidebarHTML = '';
+
+// Create one link for each Shadow Reader category.
+  for (let path in pub.categories) {
+    newSidebarHTML += template.replace('%href%', basePath + path)
+                              .replace('%title%', pub.categories[path]);
+  }
+
+  return html.replace(/(<amp-sidebar[^]+?>)[^]+<\/amp-sidebar>/, "$1\n" + newSidebarHTML + "\n</amp-sidebar>");
```

Comment:
can you explain this? AFAIK the sidebar isn't even used in the ShadowReader, so why do we have to rewrite it?

Reply:
Yes!  I did all this for the case where we simply serve a Guardian AMP article.  That article has no way to get to the Shadow Reader, so I replaced its menu with one that contains only links to the Shadow Reader.

I don't totally love this approach, but I hoped it would produce a worthy simulation of an AMP→PWA experience!

Reply:
ah nice, nifty idea! All good then :)